libnopegl/src/node_animkeyframe.c

```diff
@@ -319,6 +323,175 @@ static easing_type back_derivative(easing_type t, easing_type s)
 DECLARE_EASINGS(back,            , back_func(x, PARAM(0, 1.70158)), TRANSFORM)
 DECLARE_EASINGS(back, _derivative, back_derivative(x, PARAM(0, 1.70158)), DERIVATIVE)
 
+/* Cubic Bezier */
+
+static int double_close_to(double value, double ref, double eps)
+{
+    return fabs(value - ref) < eps;
+}
+
+static float clamp_root(float r)
+{
+    const float s = NGLI_CLAMP(r, 0.f, 1.f);
+    return fabsf(s - r) > FLT_EPSILON ? NAN : s;
+}
+
+/*
+ * Fast cbrt adapted from the Android Compose library.
+ *
+ * See: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-util/src/commonMain/kotlin/androidx/compose/ui/util/MathHelpers.kt;l=160
+ */
+static float fast_cbrt(float x)
+{
+    union { float f32; uint32_t u32; } v = { .f32 = x };
+    v.u32 = v.u32 & 0x1ffffffffL;
+    v.u32 = 0x2a510554 + (v.u32 / 3);
+
+    float estimate = v.f32;
+    estimate -= (estimate - x / (estimate * estimate)) * (1.0f / 3.0f);
+    estimate -= (estimate - x / (estimate * estimate)) * (1.0f / 3.0f);
+    return estimate;
+}
+
+/*
+ * Find the first cubic root using Cardano's algorithm.
+ *
+ * See: https://pomax.github.io/bezierinfo/#yforx
+ */
+static float find_first_cubic_root(float p0, float p1, float p2, float p3)
+{
+    double a = 3.0 * (p0 - 2.0 * p1 + p2);
+    double b = 3.0 * (p1 - p0);
+    double c = p0;
+    double d = -p0 + 3.0 * (p1 - p2) + p3;
+
+    if (double_close_to(d, 0.0, FLT_EPSILON)) {
+        if (double_close_to(a, 0.0, FLT_EPSILON)) {
+            if (double_close_to(b, 0.0, FLT_EPSILON)) {
```

Comment:
If you want to prevent a division by zero you can check directly against 0.0, it is enough to cover all `NaN` cases (assuming the numerator is finite). So I'm assuming this check is to prevent an `Inf` value. Unfortunately you can't check for that like that. For example 3e-16 will pass the epsilon check, but 1e+308/3e-16 will make an `Inf`.

Note: you probably wanted to use `DBL_EPSILON` here since you're working with double precision. But in general, epsilon is never a good threshold against `NaN`/`Inf` (and actually using double epsilon here will make it worse).